### PR TITLE
Refactor association controllers to use more common code

### PR DESCRIPTION
**Pull request by sebgl**
Fixes https://github.com/elastic/cloud-on-k8s/issues/2698.

Refactor the existing association controllers to share common code.
We now have a generic reconciler in `pkg/association`. Each association controller in `pkg/association/controller` instantiates the reconciler with its specific parameters (including association name, labels, etc.).
This removes a lot of duplicated code among the existing association controllers.

The PR also adds unit tests for the entire reconciliation logic (on top of testing each individual smaller scoped function).

### CODE REVIEW

File: pkg/controller/association/ca.go [hunk 1]
```diff
@@ -53,12 +47,12 @@ func ReconcileCASecret(
 	expectedSecret := corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: associated.GetNamespace(),
-			Name:      ElasticsearchCACertSecretName(associated, suffix),
-			Labels:    labels,
+			Name:      ElasticsearchCACertSecretName(associated, r.AssociationName),
+			Labels:    r.AssociationLabels(k8s.ExtractNamespacedName(associated)),
```
Comment (sebgl) [inline pkg/controller/association/ca.go]: we used to set the association labels + the associated resource labels here:
```yaml
"labels": {
            "apm.k8s.elastic.co/name": "kibana-sample",
            "common.k8s.elastic.co/type": "apm-server",
            "kibanaassociation.k8s.elastic.co/name": "kibana-sample",
            "kibanaassociation.k8s.elastic.co/namespace": "default"
        },
```

I think this is not really necessary, and other secrets created for an association don't have this.
I changed it to having only the association labels:

```yaml
"labels": {
            "kibanaassociation.k8s.elastic.co/name": "kibana-sample",
            "kibanaassociation.k8s.elastic.co/namespace": "default"
        },
```

File: pkg/controller/association/dynamic_watches.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+// esWatchName returns the name of the watch setup on the referenced Elasticsearch resource.
+func esWatchName(associated types.NamespacedName) string {
+	return associated.Namespace + "-" + associated.Name + "-es-watch"
+}
+
+// esUserWatchName returns the name of the watch setup on the ES user secret.
+func esUserWatchName(associated types.NamespacedName) string {
+	return associated.Namespace + "-" + associated.Name + "-es-watch"
```
Comment (pebrc) [inline pkg/controller/association/dynamic_watches.go:24]: "-user-watch"?

Comment (sebgl) [reply pkg/controller/association/dynamic_watches.go:24]: 😮  👍

Comment (pebrc) [reply pkg/controller/association/dynamic_watches.go:24]: Not that it matters much, you don't have another watch with the same name on secrets iiuc

File: pkg/controller/association/reconciler_test.go [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+	// should do nothing
+	require.NoError(t, err)
+	require.Equal(t, reconcile.Result{}, res)
+}
+
+func TestReconciler_Reconcile_DeletionTimestamp(t *testing.T) {
+	kb := sampleKibanaWithESRef()
+	now := metav1.NewTime(time.Now())
+	kb.DeletionTimestamp = &now
+	// set the pause annotation
```
Comment (pebrc) [inline pkg/controller/association/reconciler_test.go:234]: comment left over from something else?

File: pkg/controller/association/reconciler_test.go [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+func TestReconciler_Reconcile_resourceNotFound_OnDeletion(t *testing.T) {
+	// Kibana does not exist in the apiserver, but there is a leftover es user in es namespace
+	r := testReconciler(&kibanaUserInESNamespace)
+	res, err := r.Reconcile(reconcile.Request{NamespacedName: kbNamespacedName})
+	require.NoError(t, err)
+	require.Equal(t, reconcile.Result{}, res)
+	// es user secret should have been removed
+	var secret corev1.Secret
+	err = r.Client.Get(k8s.ExtractNamespacedName(&kibanaUserInESNamespace), &secret)
+	require.Error(t, err)
```
Comment (pebrc) [inline pkg/controller/association/reconciler_test.go:217]: Verify 404 error?

File: pkg/controller/association/resources.go [hunk 5]
```diff
@@ -7,19 +7,22 @@ package association
...
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// DeleteOrphanedResources deletes resources created by an association that are left over from previous reconciliation
-// attempts. Common use case is an Elasticsearch reference in Kibana or APMServer spec that was removed.
+// DeleteOrphanedResources deletes resources created by this association that are left over from previous reconciliation
+// attempts. If a user changes namespace on a vertex of an association the standard reconcile mechanism will not delete the
+// now redundant old user object/secret. This function lists all resources that don't match the current name/namespace
+// combinations and deletes them.
 func DeleteOrphanedResources(
```
Comment (pebrc) [inline pkg/controller/association/resources.go:26]: Do we need this to be exported?

File: pkg/controller/association/resources.go [hunk 6]
```diff
@@ -7,19 +7,22 @@ package association
...
-	"go.elastic.co/apm"
-	corev1 "k8s.io/api/core/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// DeleteOrphanedResources deletes resources created by an association that are left over from previous reconciliation
-// attempts. Common use case is an Elasticsearch reference in Kibana or APMServer spec that was removed.
+// DeleteOrphanedResources deletes resources created by this association that are left over from previous reconciliation
+// attempts. If a user changes namespace on a vertex of an association the standard reconcile mechanism will not delete the
```
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/association/resources.go:23]: I reused this comment from an existing function in the code, but I think the previous one was more explicit.
Changing the namespace means referencing another Elasticsearch resource with the same name, but in another namespace, I think.

File: pkg/controller/elasticsearch/certificates/remoteca/reconcile.go [hunk 7]
```diff
@@ -13,13 +13,25 @@ import (
...
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/certificates"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
-	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/remotecluster/remoteca"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
+const (
+	// TypeLabelValue is a type used to identify a Secret which contains the CA of a remote cluster.
+	TypeLabelValue = "remote-ca"
```
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/elasticsearch/certificates/remoteca/reconcile.go:25]: Yeah :(
My logic is that it's ok for association controllers to import the resource-specific packages (eg. Elasticsearch, Kibana, etc.), but not OK for the resource-specific packages to import the association controller packages.
In this logic the ES controller manipulates this label, so it needs to live in the ES controller package.

File: pkg/controller/remoteca/labels.go [hunk 8]
```diff
@@ -34,12 +32,13 @@ func remoteCAObjectMeta(
...
-			RemoteClusterNameLabelName:      remote.Name,
-			label.ClusterNameLabelName:      owner.Name,
-			common.TypeLabelName:            TypeLabelValue,
-		},
+		Labels: maps.Merge(
+			map[string]string{
+				RemoteClusterNamespaceLabelName: remote.Namespace,
+				RemoteClusterNameLabelName:      remote.Name,
+			},
+			remoteca.LabelSelector(owner.Name),
```
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/remoteca/labels.go:40]: I think I can rename this to `remoteca.Labels(owner.Name)`

File: pkg/controller/association/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+	iteration uint64
+
+	logger logr.Logger
+}
+
+// log with namespace and name fields set for the given association resource.
+func (r *Reconciler) log(associated commonv1.Associated) logr.Logger {
+	return r.logger.WithValues(
+		"namespace", associated.GetNamespace(),
+		fmt.Sprintf("%s_name", r.AssociatedShortName), associated.GetName(),
```
[the remark this answers is not in the document]
Comment (sebgl) [reply pkg/controller/association/reconciler.go:86]: What would you use instead? 
`"type", "elasticsearch", "name", "mycluster", "namespace", "ns"`?

[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/association/reconciler.go:86]: Yes I was thinking something like that and you would then query with `type: elasticsearch AND name: mycluster`

**Pull request merged**